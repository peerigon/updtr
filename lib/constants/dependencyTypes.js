"use strict";

const REGULAR = 0;
const DEV = 1;
const OPTIONAL = 2;

function fromNpm(type) {
    switch (type) {
        case "dependencies":
            return REGULAR;
        case "devDependencies":
            return DEV;
        case "optionalDependencies":
            return OPTIONAL;
        default:
            throw new Error(`Unsupported dependency type: ${ type }`);
    }
}

exports.REGULAR = REGULAR;
exports.DEV = DEV;
exports.OPTIONAL = OPTIONAL;

exports.fromNpm = fromNpm;
// yarn uses the same strings. Change this if necessary.
exports.fromYarn = fromNpm;
