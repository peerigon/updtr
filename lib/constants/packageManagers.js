"use strict";

const NPM = "npm";
const YARN = "yarn";
const SUPPORTED = [NPM, YARN];

exports.NPM = NPM;
exports.YARN = YARN;
exports.SUPPORTED = SUPPORTED;
