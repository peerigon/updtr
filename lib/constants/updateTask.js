"use strict";

const WANTED = "wanted";
const LATEST = "latest";

exports.WANTED = WANTED;
exports.LATEST = LATEST;
