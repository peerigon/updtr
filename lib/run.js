"use strict";

var async = require("async");
var childProcess = require("child_process");
var filter = require("./filter");
var EventEmitter = require("./EventEmitter.js");
var fs = require("fs");
var path = require("path");

var defaultTestCmdNpm = "npm test";
var defaultTestCmdYarn = "yarn test";

function run(config, done) {
    var emitter = new EventEmitter();
    var testCmd;
    var reporter;

    function exec(cmd, cb) {
        childProcess.exec(cmd, { maxBuffer: Infinity, encoding: "utf8", cwd: config.cwd }, cb);
    }

    function finish(err) {
        emitter.removeAllListeners();
        done(err || null);
    }

    if (typeof config.cwd !== "string") {
        throw new Error("Cannot run updtr: cwd missing");
    }

    sanitizeConfig(config);

    if (config.registry && config.useYarn) {
        throw new Error("`yarn add` does not support custom registries yet. Please use a .npmrc file to achieve this.");
    }

    reporter = config.reporter;
    testCmd = config.testCmd;

    reporter(emitter);

    emitter.emit("init", {
        cwd: config.cwd
    });
    exec(config.useYarn ? "yarn outdated --json --flat" : "npm outdated --json --long --depth=0", function (err, stdout) {
        var missing;
        var outdated;
        var infos = [];
        var tasks;

        function createTask(info, index) {
            return function (done) {
                var event = {
                    current: index + 1, // index is zero-based
                    total: tasks.length,
                    info: info,
                    testCmd: testCmd,
                    installCmd: (config.useYarn ? "yarn add" : "npm i") + (config.registry ? (" --registry " + config.registry) : "")
                };
                var testStdout;
                var installCmd;

                if (!config.useYarn && config.saveExact) {
                    event.installCmd += " --save-exact";
                }

                installCmd = event.installCmd + " " + info.name + "@" + info.updateTo + " " + info.saveCmd;

                emitter.emit("updating", event);

                async.series({
                    deleteOldVersion: async.apply(exec, (config.useYarn ? "yarn remove " : "npm remove ") + info.name + " " + info.saveCmd),
                    installNewVersion: async.apply(exec, installCmd),
                    emitTestingEvent: function (done) {
                        emitter.emit("testing", event);
                        setImmediate(done);
                    },
                    runTests: function runTests(callback) {
                        exec(testCmd, function (err, stdout, stderr) {
                            if (err) {
                                testStdout = stdout;
                                callback(err);
                                return;
                            }
                            callback(null, stdout, stderr);
                        });
                    }
                }, function (err) {
                    if (err) {
                        emitter.emit("rollback", event);
                        exec(event.installCmd + " " + info.name + "@" + info.current + " " + info.saveCmd + (config.saveExact ? " --save-exact" : ""), function (err) {
                            if (err) {
                                finish(err);
                                return;
                            }
                            emitter.emit("rollbackDone", event);
                            if (config.testStdout) {
                                event.testStdout = testStdout;
                                emitter.emit("testStdout", event);
                            }
                            done();
                        });
                        return;
                    }
                    emitter.emit("updatingDone", event);
                    done();
                });
            };
        }

        if (err && err.code > 1) {
            finish(err);
            return;
        }

        if (!stdout) {
            emitter.emit("noop");
            finish();
            return;
        }

        outdated = JSON.parse(stdout);
        if (config.useYarn) {
            if ("data" in outdated && outdated.data.body instanceof Array) {
                infos = outdated.data.body.map(function (info) {
                    return {
                        name: info[0],
                        current: info[1],
                        saveCmd: info[4] === "devDependencies" ? "--dev" : "",
                        type: info[4],
                        wanted: info[2],
                        latest: info[3],
                        updateTo: config.wanted ? info[2] : info[3]
                    };
                });
            }
        } else {
            infos = Object.keys(outdated)
            .map(function (moduleName) {
                var info = outdated[moduleName];

                info.name = moduleName;
                info.saveCmd = info.type === "devDependencies" ? "--save-dev" : "--save";
                info.updateTo = config.wanted ? info.wanted : info.latest;

                return info;
            });
        }

        infos = infos.filter(filter(config));

        missing = modulesMissing(infos);
        if (missing.length > 0) {
            emitter.emit("modulesMissing", {
                infos: missing
            });
            finish();
            return;
        }

        if (infos.length === 0) {
            emitter.emit("noop");
            finish();
            return;
        }

        emitter.emit("outdated", {
            infos: infos,
            total: infos.length
        });

        tasks = infos
            .map(createTask);

        async.series(tasks, function (err) {
            if (err) {
                finish(err);
                return;
            }
            emitter.emit("finished");
            finish();
        });
    });
}

function modulesMissing(infos) {
    return infos.filter(function (info) {
        return !info.current;
    });
}

function sanitizeConfig(config) {
    config.useYarn = fs.existsSync(path.resolve(config.cwd, "yarn.lock"));
    config.reporter = config.reporter || Function.prototype;
    config.testCmd = config.testCmd || (config.useYarn ? defaultTestCmdYarn : defaultTestCmdNpm);
    config.exclude = config.exclude && config.exclude.split(",").map(function (name) {
        return name.trim();
    }) || [];
}

module.exports = run;
