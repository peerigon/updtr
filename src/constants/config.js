// The first value in arrays is the default value

export const USE_NPM = "npm";
export const USE_YARN = "yarn";
export const USE_OPTIONS = [USE_NPM, USE_YARN];
export const UPDATE_TO_LATEST = "latest";
export const UPDATE_TO_NON_BREAKING = "non-breaking";
export const UPDATE_TO_WANTED = "wanted";
export const UPDATE_TO_OPTIONS = [
    UPDATE_TO_LATEST,
    UPDATE_TO_NON_BREAKING,
    UPDATE_TO_WANTED,
];
export const SAVE_SMART = "smart";
export const SAVE_CARET = "caret";
export const SAVE_EXACT = "exact";
export const SAVE_OPTIONS = [SAVE_SMART, SAVE_CARET, SAVE_EXACT];
