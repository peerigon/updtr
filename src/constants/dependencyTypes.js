export const REGULAR = "dependencies";
export const DEV = "devDependencies";
export const OPTIONAL = "optionalDependencies";

export function fromNpm(type) {
    switch (type) {
        case "dependencies":
            return REGULAR;
        case "devDependencies":
            return DEV;
        case "optionalDependencies":
            return OPTIONAL;
        default:
            throw new Error(`Unsupported dependency type: ${ type }`);
    }
}

export {
    // yarn uses the same strings. Change this if necessary.
    fromNpm as fromYarn,
};
