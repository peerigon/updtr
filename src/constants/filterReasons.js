export const GIT = "git";
export const EXOTIC = "exotic";
export const UNSTABLE = "unstable";
export const EXCLUDED = "excluded";
export const NOT_WANTED = "not-wanted";
