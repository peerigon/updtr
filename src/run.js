import init from "./tasks/init";
import sequentialUpdate from "./tasks/sequentialUpdate";
import splitUpdateTasks from "./tasks/util/splitUpdateTasks";
import batchUpdate from "./tasks/batchUpdate";
import createUpdateResult from "./tasks/util/createUpdateResult";
import finish from "./tasks/finish";
import updatePackageJson from "./tasks/updatePackageJson";

async function runUpdateTasks(updtr, updateTasks) {
    const results = [];
    const { breaking, nonBreaking } = splitUpdateTasks(updateTasks);
    const sequentialUpdateTasks = breaking.slice();
    let batchSuccess; // can be undefined, true or false
    let batchUpdateFailure;

    // Run batch update if we have more than one non-breaking update
    // If the batch update fails, it will roll back all modules except the first one.
    // This way we can skip one install command since we will run the sequential update for it anyway.
    if (nonBreaking.length > 1) {
        batchSuccess = await batchUpdate(updtr, nonBreaking);
    }

    if (batchSuccess === true) {
        results.push(
            ...nonBreaking.map(updateTask =>
                createUpdateResult(updateTask, true)
            )
        );
    } else {
        sequentialUpdateTasks.unshift(...nonBreaking);
        // If batchSuccess is false, we have actually executed the batch update and it returned false
        if (batchSuccess === false) {
            batchUpdateFailure = createUpdateResult(nonBreaking[0], false);
        }
    }

    // Run sequential update for all breaking updates and non-breaking batch updates that failed
    results.push(
        ...(await sequentialUpdate(
            updtr,
            sequentialUpdateTasks,
            batchUpdateFailure
        ))
    );

    return finish(updtr, results);
}

export default (async function run(updtr) {
    const results = [];

    updtr.emit("start", {
        config: updtr.config,
    });

    const { updateTasks } = await init(updtr);

    if (updateTasks.length > 0) {
        results.push(...(await runUpdateTasks(updtr, updateTasks)));
        await updatePackageJson(updtr, results);
    }

    updtr.emit("end", {
        config: updtr.config,
        results,
    });

    return results;
});
