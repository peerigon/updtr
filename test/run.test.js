import { UPDATE_TO_NON_BREAKING } from "../src/constants/config";
import run from "../src/run";
import FakeUpdtr from "./helpers/FakeUpdtr";
import pickEventNames from "./helpers/pickEventNames";
import pickEvents from "./helpers/pickEvents";
import ExecError from "./helpers/ExecError";
import {
    npmNoOutdated,
    update,
    testPass,
    testFailWithRollback,
    npmList,
} from "./fixtures/execResults";
import { outdatedRegular } from "./fixtures/packageJsons";

function npmOutdated({ nonBreaking = 0, breaking = 0 }) {
    const outdated = {};
    let i;

    for (i = 0; i < nonBreaking; i++) {
        const name = "updtr-test-module-" + (i + 1);

        outdated[name] = {
            current: "2.0.0",
            wanted: "2.1.1",
            latest: "2.1.1",
            location: "node_modules/" + name,
        };
    }
    for (i = 0; i < breaking; i++) {
        const name = "updtr-test-module-" + (i + 1);

        outdated[name] = {
            current: "1.0.0",
            wanted: "1.1.1",
            latest: "2.0.0",
            location: "node_modules/" + name,
        };
    }

    return [
        { stdout: "" }, // installMissing
        // npm exits with exit code 1 when there are outdated dependencies
        new ExecError({
            stdout: JSON.stringify(outdated),
            exitCode: 1,
        }), // outdated
    ];
}

function updateTaskNames(updateTasks) {
    return updateTasks.map(({ name }) => name);
}

function addPackageJsonFileStubs(updtr) {
    updtr.readFile.withArgs("package.json").resolves(outdatedRegular);
    updtr.writeFile.withArgs("package.json").resolves();
}

describe("run()", () => {
    it("should emit a start event of expected shape", async () => {
        const updtr = new FakeUpdtr();

        updtr.execResults = npmNoOutdated;

        await run(updtr);

        expect(updtr.emit.args.shift()).toMatchSnapshot("start event");
    });
    it("should emit init sequence events", async () => {
        const updtr = new FakeUpdtr();
        const eventNames = ["init/start", "init/end"];

        updtr.execResults = npmNoOutdated;

        await run(updtr);

        expect(pickEventNames(eventNames, updtr.emit.args)).toEqual(eventNames);
    });
    describe("when there are no outdated dependencies", () => {
        it("should not execute the sequential-update", async () => {
            const updtr = new FakeUpdtr();

            updtr.execResults = npmNoOutdated;

            await run(updtr);

            expect(
                pickEventNames(["sequential-update/start"], updtr.emit.args)
            ).toHaveLength(0);
        });
        it("should emit an end event of expected shape and return the results", async () => {
            const updtr = new FakeUpdtr();

            updtr.execResults = npmNoOutdated;

            const results = await run(updtr);
            const endEventArgs = updtr.emit.args.pop();

            expect(endEventArgs).toMatchSnapshot("no-outdated > end event");
            expect(results).toBe(endEventArgs[1].results);
        });
    });
    describe("when there is just one non-breaking update", () => {
        it("should run the sequential-update", async () => {
            const updtr = new FakeUpdtr();

            updtr.execResults = npmOutdated({ nonBreaking: 1 }).concat(
                update,
                testPass
            );
            addPackageJsonFileStubs(updtr);

            await run(updtr);

            const batchUpdateEvents = pickEvents(
                "batch-update/start",
                updtr.emit.args
            );
            const sequentialUpdateEvent = pickEvents(
                "sequential-update/start",
                updtr.emit.args
            )[0];

            expect(batchUpdateEvents).toHaveLength(0);
            expect(updateTaskNames(sequentialUpdateEvent.updateTasks)).toEqual([
                "updtr-test-module-1",
            ]);
        });
        it("should emit an end event of expected shape and return the results", async () => {
            const updtr = new FakeUpdtr();

            updtr.execResults = npmOutdated({ nonBreaking: 1 }).concat(
                update,
                testPass
            );
            addPackageJsonFileStubs(updtr);

            const results = await run(updtr);
            const endEventArgs = updtr.emit.args.pop();

            expect(endEventArgs).toMatchSnapshot("one-breaking > end event");
            expect(results).toBe(endEventArgs[1].results);
        });
    });
    describe("when there are just breaking updates", () => {
        it("should run the sequential-update for all dependencies", async () => {
            const updtr = new FakeUpdtr();

            updtr.execResults = npmOutdated({ breaking: 2 }).concat(
                update,
                testPass,
                update,
                testFailWithRollback
            );
            addPackageJsonFileStubs(updtr);

            await run(updtr);

            const batchUpdateEvents = pickEvents(
                "batch-update/start",
                updtr.emit.args
            );
            const sequentialUpdateEvent = pickEvents(
                "sequential-update/start",
                updtr.emit.args
            )[0];

            expect(batchUpdateEvents).toHaveLength(0);
            expect(updateTaskNames(sequentialUpdateEvent.updateTasks)).toEqual([
                "updtr-test-module-1",
                "updtr-test-module-2",
            ]);
        });
        it("should emit an end event of expected shape and return the results", async () => {
            const updtr = new FakeUpdtr();

            updtr.execResults = npmOutdated({ breaking: 2 }).concat(
                update,
                testPass,
                update,
                testFailWithRollback
            );
            addPackageJsonFileStubs(updtr);

            const results = await run(updtr);
            const endEventArgs = updtr.emit.args.pop();

            expect(endEventArgs).toMatchSnapshot(
                "breaking updates > end event"
            );
            expect(results).toBe(endEventArgs[1].results);
        });
    });
    describe("when there are two non-breaking updates and the update is ok", () => {
        it("should only run the batch-update", async () => {
            const updtr = new FakeUpdtr();

            updtr.execResults = npmOutdated({ nonBreaking: 2 }).concat(
                update,
                testPass
            );
            addPackageJsonFileStubs(updtr);

            await run(updtr);

            const batchUpdateEvent = pickEvents(
                "batch-update/start",
                updtr.emit.args
            )[0];
            const sequentialUpdateEvents = pickEvents(
                "sequential-update/start",
                updtr.emit.args
            );

            expect(updateTaskNames(batchUpdateEvent.updateTasks)).toEqual([
                "updtr-test-module-1",
                "updtr-test-module-2",
            ]);
            expect(sequentialUpdateEvents).toHaveLength(0);
        });
        it("should emit an end event of expected shape and return the results", async () => {
            const updtr = new FakeUpdtr();

            updtr.execResults = npmOutdated({ nonBreaking: 2 }).concat(
                update,
                testPass
            );
            addPackageJsonFileStubs(updtr);

            const results = await run(updtr);
            const endEventArgs = updtr.emit.args.pop();

            expect(endEventArgs).toMatchSnapshot(
                "two non-breaking and update ok > end event"
            );
            expect(results).toBe(endEventArgs[1].results);
        });
    });
    describe("when there are two non-breaking updates and the update is not ok", () => {
        it("should run the batch-update first and back off to the sequential-update", async () => {
            const updtr = new FakeUpdtr();

            updtr.execResults = npmOutdated({ nonBreaking: 2 }).concat(
                update,
                testFailWithRollback,
                testFailWithRollback,
                testPass
            );
            addPackageJsonFileStubs(updtr);

            await run(updtr);

            const batchUpdateEvent = pickEvents(
                "batch-update/start",
                updtr.emit.args
            )[0];
            const sequentialUpdateEvent = pickEvents(
                "sequential-update/start",
                updtr.emit.args
            )[0];

            expect(updateTaskNames(batchUpdateEvent.updateTasks)).toEqual([
                "updtr-test-module-1",
                "updtr-test-module-2",
            ]);
            expect(updateTaskNames(sequentialUpdateEvent.updateTasks)).toEqual([
                "updtr-test-module-1",
                "updtr-test-module-2",
            ]);
        });
        it("should emit an end event of expected shape and return the results", async () => {
            const updtr = new FakeUpdtr();

            updtr.execResults = npmOutdated({ nonBreaking: 2 }).concat(
                update,
                testFailWithRollback,
                testFailWithRollback,
                testPass
            );
            addPackageJsonFileStubs(updtr);

            const results = await run(updtr);
            const endEventArgs = updtr.emit.args.pop();

            expect(endEventArgs).toMatchSnapshot(
                "two non-breaking and update not ok > end event"
            );
            expect(results).toBe(endEventArgs[1].results);
        });
    });
    describe(`when updateTo is "${ UPDATE_TO_NON_BREAKING }"`, () => {
        it("should finish incomplete results", async () => {
            const updtr = new FakeUpdtr({
                updateTo: UPDATE_TO_NON_BREAKING,
            });

            // npm outdated reports a breaking change, but the updateTo option
            // limits the update to a non-breaking version
            updtr.execResults = npmOutdated({ breaking: 1 }).concat(
                update,
                testPass,
                npmList
            );
            addPackageJsonFileStubs(updtr);

            expect(await run(updtr)).toMatchSnapshot(
                `updateTo ${ UPDATE_TO_NON_BREAKING } > end event`
            );
        });
    });
});
