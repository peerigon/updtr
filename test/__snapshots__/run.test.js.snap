// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`breaking updates > end event 1`] = `
Array [
  "end",
  Object {
    "config": Object {
      "cwd": "/updtr/test/cwd",
      "exclude": Array [],
      "registry": undefined,
      "save": "smart",
      "test": undefined,
      "updateTo": "latest",
      "use": "npm",
    },
    "results": Array [
      Object {
        "name": "updtr-test-module-1",
        "rollbackTo": "1.0.0",
        "success": true,
        "updateTo": "2.0.0",
      },
      Object {
        "name": "updtr-test-module-2",
        "rollbackTo": "1.0.0",
        "success": false,
        "updateTo": "2.0.0",
      },
    ],
  },
]
`;

exports[`no-outdated > end event 1`] = `
Array [
  "end",
  Object {
    "config": Object {
      "cwd": "/updtr/test/cwd",
      "exclude": Array [],
      "registry": undefined,
      "save": "smart",
      "test": undefined,
      "updateTo": "latest",
      "use": "npm",
    },
    "results": Array [],
  },
]
`;

exports[`one-breaking > end event 1`] = `
Array [
  "end",
  Object {
    "config": Object {
      "cwd": "/updtr/test/cwd",
      "exclude": Array [],
      "registry": undefined,
      "save": "smart",
      "test": undefined,
      "updateTo": "latest",
      "use": "npm",
    },
    "results": Array [
      Object {
        "name": "updtr-test-module-1",
        "rollbackTo": "2.0.0",
        "success": true,
        "updateTo": "2.1.1",
      },
    ],
  },
]
`;

exports[`start event 1`] = `
Array [
  "start",
  Object {
    "config": Object {
      "cwd": "/updtr/test/cwd",
      "exclude": Array [],
      "registry": undefined,
      "save": "smart",
      "test": undefined,
      "updateTo": "latest",
      "use": "npm",
    },
  },
]
`;

exports[`two non-breaking and update not ok > end event 1`] = `
Array [
  "end",
  Object {
    "config": Object {
      "cwd": "/updtr/test/cwd",
      "exclude": Array [],
      "registry": undefined,
      "save": "smart",
      "test": undefined,
      "updateTo": "latest",
      "use": "npm",
    },
    "results": Array [
      Object {
        "name": "updtr-test-module-1",
        "rollbackTo": "2.0.0",
        "success": false,
        "updateTo": "2.1.1",
      },
      Object {
        "name": "updtr-test-module-2",
        "rollbackTo": "2.0.0",
        "success": true,
        "updateTo": "2.1.1",
      },
    ],
  },
]
`;

exports[`two non-breaking and update ok > end event 1`] = `
Array [
  "end",
  Object {
    "config": Object {
      "cwd": "/updtr/test/cwd",
      "exclude": Array [],
      "registry": undefined,
      "save": "smart",
      "test": undefined,
      "updateTo": "latest",
      "use": "npm",
    },
    "results": Array [
      Object {
        "name": "updtr-test-module-1",
        "rollbackTo": "2.0.0",
        "success": true,
        "updateTo": "2.1.1",
      },
      Object {
        "name": "updtr-test-module-2",
        "rollbackTo": "2.0.0",
        "success": true,
        "updateTo": "2.1.1",
      },
    ],
  },
]
`;

exports[`updateTo non-breaking > end event 1`] = `
Array [
  Object {
    "name": "updtr-test-module-1",
    "rollbackTo": "1.0.0",
    "success": true,
    "updateTo": "2.0.0",
  },
]
`;
